Sherlock/replicator.h

```diff
@@ -0,0 +1,285 @@
+      ++index_;
+      event_.notify_one();
+    }
+
+    void TerminateSubscription() {
+      std::unique_lock<std::mutex> lock(mutex_);
+      event_.wait(lock, [this]() { return !subscription_id_.empty(); });
+      external_stop_ = true;
+      const std::string terminate_url =
+          remote_stream_.ObjectAccessorDespitePossiblyDestructing().url_ + "?terminate=" + subscription_id_;
+      try {
+        const auto result = HTTP(GET(terminate_url));
+      } catch (current::net::NetworkException&) {
+      }
+    }
+  };
+
+  template <typename F, typename TYPE_SUBSCRIBED_TO>
+  class RemoteSubscriberScope final : public current::sherlock::SubscriberScope {
+   private:
+    static_assert(current::ss::IsStreamSubscriber<F, TYPE_SUBSCRIBED_TO>::value, "");
+    using base_t = current::sherlock::SubscriberScope;
+
+   public:
+    using subscriber_thread_t = RemoteSubscriberThread<F, TYPE_SUBSCRIBED_TO>;
+
+    RemoteSubscriberScope(ScopeOwned<RemoteStream>& remote_stream,
+                          F& subscriber,
+                          uint64_t start_idx,
+                          std::function<void()> done_callback)
+        : base_t(std::move(
+              std::make_unique<subscriber_thread_t>(remote_stream, subscriber, start_idx, done_callback))) {}
+  };
+
+  explicit SubscribableRemoteStream(const std::string& remote_stream_url)
+      : stream_(remote_stream_url),
+        schema_(
+            Value<reflection::ReflectedTypeBase>(reflection::Reflector().ReflectType<T_STREAM_ENTRY>()).type_id,
+            sherlock::constants::kDefaultTopLevelName,
+            sherlock::constants::kDefaultNamespaceName) {
+    CheckRemoteSchema();
+  }
+
+  explicit SubscribableRemoteStream(const std::string& remote_stream_url,
+                                    const std::string& top_level_name,
+                                    const std::string& namespace_name)
+      : stream_(remote_stream_url),
+        schema_(
+            Value<reflection::ReflectedTypeBase>(reflection::Reflector().ReflectType<T_STREAM_ENTRY>()).type_id,
+            top_level_name,
+            namespace_name) {
+    CheckRemoteSchema();
+  }
+  ~SubscribableRemoteStream() {}
+
+  template <typename F>
+  RemoteSubscriberScope<F, T_STREAM_ENTRY> Subscribe(F& subscriber,
+                                                     uint64_t start_idx = 0u,
+                                                     std::function<void()> done_callback = nullptr) {
+    static_assert(current::ss::IsStreamSubscriber<F, T_STREAM_ENTRY>::value, "");
+    try {
+      return RemoteSubscriberScope<F, T_STREAM_ENTRY>(stream_, subscriber, start_idx, done_callback);
+    } catch (const current::sync::InDestructingModeException&) {
+      CURRENT_THROW(StreamInGracefulShutdownException());
+    }
+  }
+
+ private:
+  void CheckRemoteSchema() {
+    const auto response = HTTP(GET(stream_.ObjectAccessorDespitePossiblyDestructing().url_ + "/schema.simple"));
```

Comment:
Design suggestion for outside this pull request. Perhaps it's best to have a single "pre-subscribe" endpoint, would would:
1) Accept the desired schema as a header.
2) If the actual schema matches the desired one, pre-generate a subscription ID and return it, so that the actual subscription that follows next can request using the pre-generated ID instead of waiting for one in the header.
3) If the actual schema does not match, return both the 3-tuple schema signature and the actual schema, so that if we need to know what exactly is the difference between the schemas, the data to diff is right there.